### PR TITLE
Add Styleguide 

### PR DESCRIPTION
**Pull request by Mab879**
#### Description:

I have created a very rough draft of a styleguide. This PR will need lots of feedback. I am really looking for advise on the remediation and OVAL sections.  

#### Rationale:

To help give our project guide of how we want content to be laid out.

### CODE REVIEW

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:6]: Maybe some statement here about trailing white space in general?

"When not syntactically necessary, ..."

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:12]: Might be nice to clarify here that we prefer Py2+Py3 compatible files, though when necessary, utilities can be Python3-only (see e.g. https://github.com/ComplianceAsCode/content/pull/7454#discussion_r697872005)

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:14]: Interestingly, YAML forbids tabs and prefers 2 spaces: https://docs.saltproject.io/en/latest/topics/troubleshooting/yaml_idiosyncrasies.html#spaces-vs-tabs 

Thoughts? Though, admittedly there isn't a better source than this that I can find easily.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:17]: I'd just add a semicolon here at the end. Just a nit.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:21]: Maybe clarify that these are Rule YAMLs (and live with duplicating YAML from the parent section).

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:25]: I think we can elide the values and point to the documentation on Rule YAML's elements if you want.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:23]: I'd leave out the "required"-ness as I'm 80% sure that's covered elsewhere (where the _format_ is described) and if not, should be added there. ;-)

Comment (cipherboy) [reply docs/manual/developer/04_styleguide.md:23]: If you wanted, you could probably say something along the lines of "in the following order, if present".

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:34]: I think we should mention here this includes some (subset of?) HTML here and doesn't support Markdown.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:35]: As above about HTML-like.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:37]: Drop this as above?

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:42]: Keys must be in alphabetical order. Some keys may have multi-part comma-separated values, but here we don't require a sorted order there.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:44]: As above about identifiers.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:45]: I'd prefer `platforms` and leave the reference to `platform` as a legacy note rather than putting it first. My 2c.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:46]: Is this explicitly present anywhere?

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:50]: I think this is as above about HTML-like.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:69]: As above about HTML-like.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [reply docs/manual/developer/04_styleguide.md:23]: This also applies below to Groups &c.

File: docs/manual/developer/04_styleguide.md
Comment (cipherboy) [inline docs/manual/developer/04_styleguide.md:60]: Oddly, Group has no sub-header called "sections" but Rules does. Should we duplicate the structure here?

**Review by cipherboy — COMMENTED** [review]
@Mab879 had asked for feedback and thoughts from me, I realize this is an early draft but figured I'd provide some now while I'm thinking about other things to include.

File: docs/manual/developer/04_styleguide.md
Comment (Mab879) [reply docs/manual/developer/04_styleguide.md:14]: We are inconstant across the codebase. We need to pick one and stick with it. I'm fine with two if that the way we want to go.

File: docs/manual/developer/04_styleguide.md
Comment (Mab879) [reply docs/manual/developer/04_styleguide.md]: No, removed.

File: docs/manual/developer/04_styleguide.md
Comment (jan-cerny) [reply docs/manual/developer/04_styleguide.md]: Great point! We discussed this internally couple of weeks ago and we prefer switching to Python 3. Python 2 was needed due to RHEL7 and similar, but starting from RHEL 7.7 there is Python 3.6 in RHEL 7, so  Python 2 isn't needed anymore.

Comment (matejak) [reply docs/manual/developer/04_styleguide.md]: Not so fast, we haven't made the switch yet, so build-related and test-related files should be compatible for the time being. I hope that we will be able to migrate to 3.6 this calendar year.

File: docs/manual/developer/04_styleguide.md
Comment (matejak) [reply docs/manual/developer/04_styleguide.md]: Let's pick a style that is least disruptive - you don't have to have the history cluttered by reformatting when you are trying to find out what the historical reasons of why the code looks in some particular way.

File: docs/manual/developer/04_style_guide.md
Comment (matejak) [inline docs/manual/developer/04_style_guide.md:170]:
```suggestion
* [One sentence per line](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line)
```

File: docs/manual/developer/04_style_guide.md
Comment (matejak) [inline docs/manual/developer/04_style_guide.md:132]:
```suggestion
* Use 4-space indentation
* Put `do` or `then` on the same line as `for` or `if` respectively, e.g. `for file in *; do`

```

File: docs/manual/developer/04_style_guide.md
Comment (matejak) [inline docs/manual/developer/04_style_guide.md:9]:
```suggestion
* The milestone must be set correctly.
```

File: docs/manual/developer/04_style_guide.md
Comment (matejak) [inline docs/manual/developer/04_style_guide.md:12]:
```suggestion
* There are no merge commits - if that's the case , they have to be taken out by rebasing.

```

File: docs/manual/developer/04_styleguide.md
Comment (Mab879) [reply docs/manual/developer/04_styleguide.md]: I will keep `The files in the build system shall be Python 2.7 and Python 3 compatible` in the Python section then.

File: docs/manual/developer/04_style_guide.md
Comment (yuumasato) [inline docs/manual/developer/04_style_guide.md:18]: Did you mean a newline at the end of the file, instead of a blank line?

File: docs/manual/developer/04_style_guide.md
Comment (yuumasato) [inline docs/manual/developer/04_style_guide.md:65]:
```suggestion
  * Must be one line 
```

File: docs/manual/developer/04_style_guide.md
Comment (yuumasato) [inline docs/manual/developer/04_style_guide.md:99]: A group can have `platforms` key too.

File: docs/manual/developer/04_style_guide.md
Comment (yuumasato) [inline docs/manual/developer/04_style_guide.md:122]:
```suggestion
All remediations should have the following header with the appropriate values
```

File: docs/manual/developer/04_style_guide.md
Comment (yuumasato) [inline docs/manual/developer/04_style_guide.md:132]:
```suggestion
* Should use macros instead shared functions
```

File: docs/manual/developer/04_style_guide.md
Comment (yuumasato) [inline docs/manual/developer/04_style_guide.md:176]: Typo, I guess?
```suggestion
* Must have one blank line above headings
```

File: docs/manual/developer/04_style_guide.md
Comment (yuumasato) [inline docs/manual/developer/04_style_guide.md:177]: Should this apply to all heading levels?
I see, for example,  in line 165 `### State elements`, should it be `### State Elements`?

File: docs/manual/developer/04_style_guide.md
Comment (jan-cerny) [inline docs/manual/developer/04_style_guide.md:220]:
```suggestion
* Shall use the `.sh` for the file extension
* Should use the `#!/bin/bash` shebang at the first line
```

File: docs/manual/developer/04_styleguide.md
Comment (Mab879) [reply docs/manual/developer/04_styleguide.md]: I added a note about new files being four and existing files can be two spaces.

File: docs/manual/developer/04_style_guide.md
Comment (jan-cerny) [inline docs/manual/developer/04_style_guide.md:225]: The link is for asciidoc which is a different format than markdown

Comment (Mab879) [reply docs/manual/developer/04_style_guide.md:225]: This is true, but the ideas are similar @matejak suggested this link.

File: docs/manual/developer/04_style_guide.md
Comment (cipherboy) [inline docs/manual/developer/04_style_guide.md:208]: Should these be indented to clarify they're part of OVAL?

File: docs/manual/developer/04_style_guide.md
Comment (cipherboy) [inline docs/manual/developer/04_style_guide.md:194]: Do we want to add something about OVAL comments here? IMO they're really helpful and while mandated in certain places, aren't mandated everywhere and thus we should encourage their use and clarity?

File: docs/manual/developer/04_style_guide.md
Comment (cipherboy) [inline docs/manual/developer/04_style_guide.md:101]: nit: Double star I think is a typo?

**Review by cipherboy — COMMENTED** [review]
General thoughts from me are that I'd like to see a few more linked resources in these documents, perhaps to justify and/or clarify some of these decisions. E.g., `git rebase` discussion is justified in text, but the _how_ of performing this action is, IMO, omitted, where we have the space to link to some best practice guide.

My 2c., but like I just hinted at above, perhaps this document is best called a "best practice" guide rather than strictly a style guide? 

Additionally, JINJA macros aren't discussed anywhere -- IMO there are/should be best practices around them (such as documenting them and adding doc headers &c). 

Might also be worth calling out when a rule/... should be templated? I don't see any discussion on shared/templates.

**Edit**: And thanks to Matt and everyone for doing this! Looking very good, especially compared to the first draft I saw :-)

File: docs/manual/developer/04_style_guide.md
Comment (Mab879) [reply docs/manual/developer/04_style_guide.md:11]: Reason why I put in the Before Merging section. My goal of this section is to be checklist for the maintainer merging the PR. Maybe I need to add something to make it clearer.

**Review by yuumasato — APPROVED** [review]
Thank you the guide, :)

**Pull request merged**
